### ISSUE TITLE
Changing default behavior of module.to() to better support complex-valued parameters

### ISSUE THREAD

**Issue by allevitan** [pytorch/pytorch#95337]
### 🚀 The feature, motivation and pitch

Motivation
--------------

It is incredibly common for modules which use complex-valued parameters to also contain real-valued parameters. This causes problems wherever module.to() is called.

Consider the example below:

```python
real = torch.ones([10,10], dtype=torch.float32)

# returns torch.complex64
print(real.to(dtype=torch.complex64).dtype)

complex = torch.ones([10,10], dtype=torch.complex64)

# returns torch.float32 and raises UserWarning for casting complex to real
print(complex.to(dtype=torch.float32).dtype)
```

This behavior seems obviously correct, but a problem appears at the module level:

```python
class MixedModule(torch.nn.Module):
     def __init__(self):
         super().__init__()
         self.complex_param = torch.nn.Parameter(torch.ones([10,10], dtype=torch.complex64))
         self.real_param = torch.nn.Parameter(torch.ones([10,10], dtype=torch.float32))

mm = MixedModule()

mm.to(dtype=torch.float64) 

 # returns torch.float64
print(mm.real_param.dtype)

# returns torch.float64
print(mm.complex_param.dtype)
```

This is makes it a dangerous game to use `module.to()` on any module which has mixed real and complex-valued parameters, for example to ensure that all parameters loaded from some external data source are stored at the proper precision for the module to be run. 

This issue also exists when modules have mixed integer and floating-point variables, which is why [module.to](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.to) only accepts floating-point values, and will not update the dtype of any integer variables.

Proposed Feature
-----------------------

I propose add a new kwarg, `precision`, to both `Tensor.to` and `Module.to`, which will update the precision but not affect the realness or complex-ness of the tensor(s) to be updated. Here is a brief proposal for it's behavior:

1) It accepts as input a `t.dtype`, whose precision it will attempt to match, e.g.:
```python
real = torch.ones([10,10], dtype=torch.float32)

# returns torch.float32
print(real.to(precision=torch.complex64).dtype)

# returns torch.float64
print(real.to(precision=torch.float64).dtype)

# returns torch.float64
print(real.to(precision=torch.complex128).dtype)
```

2) If both a dtype and precision are input, it chooses dtype over precision and raises a UserWarning

3) It raises a TypeError when asked to convert a complex tensor to a non-supported real dtype, e.g.:
```python
complex = torch.ones([10,10], dtype=torch.complex64)

# raises TypeError
complex.to(precision=torch.float16)
```

4) In all other respects, it acts just like the dtype argument. For example, in `Tensor.to`, it would also accept integer dtypes and happily operate on integer dtypes. In `Module.to`, it would treat integers just like they are treated by the `dtype` argument in `Module.to`. The value of this choice is that it enables other code which wants the "precision"-style behavior to make one change everywhere.

Notes
--------

It is possible that other code within pytorch which uses `Module.to` to update dtypes would want to have it's default behavior changed to use the `precision` kwarg, in which case this feature addition would require making backwards-incompatible changes to those functions.

### Alternatives

Alternative 1
----------------
Instead of introducing a `precision` kwarg to `Tensor.to` and `Module.to`, we could introduce a boolean `preserve_realness` kwarg which would switch the `dtype` kwarg to have the behaviour outlined above. Is there a better name for this kwarg? In this case, ideally the default setting for `Tensor.to` would be `False` but the default setting for `Module.to` would be `True`.

The advantage is that other functions which use `Module.to` do not need to be explicitly updated, but the disadvantage is that `Module.to` itself becomes explicitly backwards-incompatible.

Alternative 2
----------------
Introduce two new functions, `Tensor.change_precision` and `Module.change_precision`. I think this is much less clean, because pytorch already chooses to combine device changes and dtype changes into one function - why introduce a new function just for this one weird thing?

Alternative 3
----------------
The exact behavior of the `precision` kwarg could also be different in a few ways:

1) It could accept integer (e.g. 32, 64) or string (e.g. 'float', 'double') input, or multiple different input styles
2) It could only accept `t.float32`, `t.float64`, `t.complex64`, and `t.complex128`, rather than replicating the full functionality of the `dtype` arg.
3) It could raise an error instead of a warning if both a dtype and precision argument are input

### Additional context

Let's all remember that complex numbers and real numbers are friends!!!

**Comment by allevitan** [comment]
Doing a further search after clarifying my thinking on this, I noted that there was some discussion [here](https://github.com/pytorch/pytorch/issues/46374#issuecomment-1383220284) noting that this was a challenge.

**Comment by ezyang** [comment]
@allevitan I have a counterproposal. Let's just retroactively redefine `float()/double()` to "do the right thing" with complex parameters (convert them to their complex equivalents). No new kwarg. How does that sound? If you're up for implementing it I will review.

**Comment by allevitan** [comment]
That sounds good! I am up to implement. Here are a few points for discussion:

1) Whatever else happens, `.float()` and `.double()` should be updated to "do the right thing".
2) I think there **also** should be a way to get the desired behavior using `.to()`. I tend to prefer using `.to()` because it lets me set the desired dtype as a parameter somewhere else. Compare:

```python
my_precision=t.float32

# Using .to()
m.to(precision=my_precision)

# using .float()/.double(): clunky, and needs to be updated if other complex precisions become available
if my_precision == t.float32:
    m.float()
else:
    m.double()
```

3) **Are there any other functions which deserve to be updated at the same time?** Currently, It's just`Tensor.float()/double()` and `Module.float()/double()`?

4) What should happen to `.cfloat()` and `cdouble()`? If `float()` and `double()` are updated, then the behavior of `cdouble()` and `cfloat()` will be confusingly different.

4) How should backwards incompatibility warnings be handled? Here is my proposal:
  - A warning is thrown when .float() or .double() is called on a complex parameter, noting the impending change and suggesting that users switch to `.to(dtype=t.float32)', etc. to preserve the old behavior
  - a `.. warning::` is added to the docs for both programs, noting the impending change (not `.. deprecated::`, because the functions will continue to exist?)
  - Either add a new kwarg to `.float()` and `.double()` to let users get the new behavior faster, or don't and just force everyone to wait but keep the code base cleaner.

**Comment by lezcano** [comment]
The solution you are proposing, Ed, seems a bit dangerous and counterintuitive to me. 
Note that if you run the script in the OP, PyTorch already  warns you that you're doing something fishy:
```
/usr/lib/python3.10/site-packages/torch/nn/modules/module.py:987: UserWarning: Casting complex values to real discards the imaginary part (Triggered internally at /build/python-pytorch/src/pytorch-1.13.1/aten/src/ATen/native/Copy.cpp:250.)
  return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
```
In this case, this can be implemented via `apply` in a rather straightforward way.

If we want to create a shorthand for this `apply`, I rather prefer the `precision` kwarg than doing something funny with `float/double`. cc @albanD

**Comment by ezyang** [comment]
@lezcano I just fail to see a situation where I have a PyTorch nn module with complex float parameters, I call double() on it, and I want it truncated to real with double precision parameters. If this is never the right thing to do, let's do the righ tthing instead.

**Comment by lezcano** [comment]
My point is more along the lines of how to have methods `float/double/cfloat/cdouble` (and their `to` counterparts) that are consistent with each other. I guess that you'd propose to have `float/cfloat` and `double/cdouble` do the same thing when the Module has real and complex tensors. But what would be the semantics of calling `cfloat()` (or `cdouble()`) on a Module that has only real tensors? 

What about the following variation:
Let's have `float/double` just act on real tensors and `cfloat/cdouble` just act on complex tensors. We could then raise a warning when the Module has real and complex tensors (or not say anything and simply document this)

**Comment by allevitan** [comment]
@lezcano I see why changing the behavior of `float` and `double` would make things inconsistent, both with `cfloat` and `cdouble` and between `module.float` and `tensor.float`. If we changed the behavior to what @ezyang suggested, it would probably be cleanest to remove `cfloat` and `cdouble` entirely, but that creates even more breaking changes...

However, I think the behavior you're proposing is probably the same as / worse than the status quo, because their **should*** be a way to ask a module to convert all floating point & complex variables to a compatible level of precision (float32/complex64 or float64/complex128). @lezcano your proposal still prevent that from happening

**Comment by lezcano** [comment]
With my proposal you'd write `model.double().cdouble()` to achieve this.

**Comment by ezyang** [comment]
That seems worse. Like, do the "intuitive" thing (double "does what you mean" when applied to complex tensors), or do the "consistent" thing (double does the same thing tensor.double would do, warts and all); this weird halfway house which is not intuitive and not consistent seems like the worst of both worlds.

**Comment by lezcano** [comment]
Note that we currently do this with int vs floats, so my proposed semantics would be consistent with this
```python
import torch

class MixedModule(torch.nn.Module):
     def __init__(self):
         super().__init__()
         self.complex_param = torch.nn.Parameter(torch.ones([10,10], dtype=torch.complex64))
         self.real_param = torch.nn.Parameter(torch.ones([10,10], dtype=torch.float32))
         self.register_buffer("int_buff", torch.ones([10,10], dtype=torch.int32))
         self.register_buffer("float_buff", torch.ones([10,10], dtype=torch.float32))

mm = MixedModule()
mm.to(dtype=torch.float64)
 # returns torch.float64
print(mm.real_param.dtype)
# returns torch.float64
print(mm.complex_param.dtype)
# returns torch.int32
print(mm.int_buff.dtype)
# returns torch.float64
print(mm.float_buff.dtype)

mm = MixedModule()
mm.to(dtype=torch.cdouble)
 # returns torch.complex128
print(mm.real_param.dtype)
# returns torch.complex128
print(mm.complex_param.dtype)
# returns torch.int32
print(mm.int_buff.dtype)
# returns torch.complex128
print(mm.float_buff.dtype)
```

**Comment by ezyang** [comment]
OK, fine, I'll buy it. In particular, 

```
>>> torch.complex64.is_floating_point
False
```

So this in fact the behavior we have today.

@allevitan so the claim is, if you want to go to double precision, instead of using to(), use `double().cdouble()` to convert both real and complex parameters/buffers. Does that satisfy your use case? We can document it if it does.

**Comment by allevitan** [comment]
Hmm, I have to admit I don't find the solution above to be that compelling. Let me first say how I understood it, in case I understood wrong:

- `module.double()` (+ `float`) currently will only update real-valued floats to double-precision. We preserve that behavior
- `module.cdouble()` (+`cfloat`) doesn't exist at the moment, but will be created to follow the same behavior for complex-valued floats
- `module.to` remains unchanged (module.to will map complex floats to real floats happily). See the source, e.g. for [module.double](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.double) vs [module.to](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.to).

This does provide a way to do what I want to, but it still feels pretty inconsistent and confusing, for a few reasons:

1) For purely real-valued models, `module.double()` is equivalent to `module.to(dtype=t.float64)`, but this change causes them to act differently for complex-valued models (but not for complex-valued tensors!)
2) Like @ezyang noted, I've never wanted to map all the real-valued floats to float64, but keep the complex floats as they are.  For a baseline, think about how things worked before complex number support was implemented. Back then, complex numbers were floating point tensors with final dimension 2. The behavior for conversion between float32 and float64 was actually more sane, because "complex-ness" was not considered a part of the dtype of a tensor.


Thinking about all this discussion, I really am drawn back to my first proposal, because it creates a simple way to do the "right" thing

An alternate or add-on proposal could be to introduce functions `module.single_precision`, `module.double_precision`, `tensor.single_precision` and `tensor.double_precision`, which work the way @ezyang suggested in his first comment, affecting all tensors. 

But, this isn't a hill I feel like dying on, I'm happy if it just becomes possible to do this conversion easily!

**Comment by ezyang** [comment]
Let's just add a new function, along the lines of what you suggested. We will need to bikeshed the name

**Comment by lezcano** [comment]
whatever works really, I don't mind the extra function, but yeah, let's think of a prettier name. 
But FWIW, as I said in https://github.com/pytorch/pytorch/issues/95337#issuecomment-1447770596, my proposal was to also modify the `to()` behaviour accordingly for consistency, but yeah.

**Comment by ezyang** [comment]
someone should ask chat gpt for a name lol
